File: backend/middleware/auth_middleware.py
import datetime
from functools import wraps
from flask import request, jsonify, g
from backend.config import Config

try:
    import jwt
    JWT_AVAILABLE = True
except ImportError:
    import base64, json as _json
    JWT_AVAILABLE = False

def create_token(user_id, rol):
    payload = {
        "sub": user_id,
        "rol": rol,
        "iat": int(datetime.datetime.utcnow().timestamp()),
        "exp": int((datetime.datetime.utcnow() + datetime.timedelta(seconds=Config.TOKEN_EXPIRY)).timestamp()),
    }
    if JWT_AVAILABLE:
        return jwt.encode(payload, Config.SECRET_KEY, algorithm="HS256")
    import base64, json
    return base64.urlsafe_b64encode(json.dumps(payload).encode()).decode()

def decode_token(token):
    if JWT_AVAILABLE:
        import jwt as _jwt
        return _jwt.decode(token, Config.SECRET_KEY, algorithms=["HS256"])
    import base64, json
    return json.loads(base64.urlsafe_b64decode(token.encode() + b"=="))

def require_auth(f):
    @wraps(f)
    def wrapper(*args, **kwargs):
        auth = request.headers.get("Authorization", "")
        if not auth.startswith("Bearer "):
            return jsonify({"error": "No autorizado"}), 401
        try:
            g.current_user = decode_token(auth[7:])
        except Exception:
            return jsonify({"error": "Token inválido o expirado"}), 401
        return f(*args, **kwargs)
    return wrapper

def require_admin(f):
    @wraps(f)
    def wrapper(*args, **kwargs):
        auth = request.headers.get("Authorization", "")
        if not auth.startswith("Bearer "):
            return jsonify({"error": "No autorizado"}), 401
        try:
            payload = decode_token(auth[7:])
            g.current_user = payload
        except Exception:
            return jsonify({"error": "Token inválido o expirado"}), 401
        if payload.get("rol") != "admin":
            return jsonify({"error": "Acceso denegado"}), 403
        return f(*args, **kwargs)
    return wrapper

File: backend/src/middleware/authMiddleware.js
const jwt = require('jsonwebtoken');
const dotenv = require('dotenv');

dotenv.config();

const JWT_SECRET = process.env.JWT_SECRET || 'pandea-secret-key';

function requireAuth(req, res, next) {
  const authHeader = req.headers.authorization || '';
  if (!authHeader.startsWith('Bearer ')) {
    return res.status(401).json({ error: 'No autorizado' });
  }
  const token = authHeader.split(' ')[1];
  try {
    const payload = jwt.verify(token, JWT_SECRET);
    req.user = payload;
    next();
  } catch (error) {
    return res.status(401).json({ error: 'Token inválido o expirado' });
  }
}

function requireAdmin(req, res, next) {
  requireAuth(req, res, () => {
    if (req.user.rol !== 'admin') {
      return res.status(403).json({ error: 'Acceso denegado' });
    }
    next();
  });
}

module.exports = {
  requireAuth,
  requireAdmin,
};
